clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -164,25 +164,29 @@ public void abort(RuntimeException exception) {
      * @param exception The exception that occurred (or null if the request was successful)
      */
     public void done(long baseOffset, long logAppendTime, RuntimeException exception) {
-        final FinalState finalState;
-        if (exception == null) {
+        final FinalState tryFinalState = (exception == null) ? FinalState.SUCCEEDED : FinalState.FAILED;
```

Comment:
It is probably cleaner to have an explicit `EXPIRED` state.

Reply:
I did some digging around. An expired batch's final state is `FAILED`. I don't feel great about adding yet another `finalState`. We already have `ABORTED` and `FAILED`.